include/picongpu/particles/startPosition/OnePositionImpl.hpp

```diff
@@ -56,19 +90,37 @@ namespace acc
         )
         {
             particle[ position_ ] = T_ParamClass{}.inCellOffset.template shrink< simDim >( );
-            particle[ weighting_ ] = m_weighting;
+
+            // set the weighting attribute if the particle species has it
+            bool const hasWeighting = pmacc::traits::HasIdentifier<
+                typename T_Particle::FrameType,
+                weighting
+            >::type::value;
+            detail::SetWeighting< hasWeighting > setWeighting;
+            setWeighting( particle, m_weighting );
         }
 
+        template< typename T_Particle >
         DINLINE uint32_t
         numberOfMacroParticles( float_X const realParticlesPerCell )
         {
-            return detail::WeightMacroParticles{}(
-                realParticlesPerCell,
-                T_ParamClass::numParticlesPerCell,
-                m_weighting
-            );
+            bool const hasWeighting = pmacc::traits::HasIdentifier<
+                typename T_Particle::FrameType,
+                weighting
+            >::type::value;
+
+            if( hasWeighting )
```

Comment:
Could you please transform this if that we have only one exit point within the method. This is a better pattern for compiler to optimize the code.

```C++

uint32_t result = T_ParamClass::numParticlesPerCell;
if( hasWeighting )
    result = startPosition::detail::WeightMacroParticles{}(
       realParticlesPerCell,
       T_ParamClass::numParticlesPerCell,
       m_weighting
    );
return result;
```